node/src/types/value_or_chunk.rs

```diff
@@ -0,0 +1,172 @@
+use std::fmt::{self, Debug, Display, Formatter};
+
+use casper_execution_engine::storage::trie::TrieRaw;
+use casper_types::bytesrepr::{self, ToBytes};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use casper_hashing::{ChunkWithProof, Digest, MerkleConstructionError};
+use datasize::DataSize;
+
+/// Represents a value or a chunk of data with attached proof.
+///
+/// Chunk with attached proof is used when the requested
+/// value is larger than [ChunkWithProof::CHUNK_SIZE_BYTES].
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub enum ValueOrChunk<V> {
+    /// Represents a value.
+    Value(V),
+    /// Represents a chunk of data with attached proof.
+    ChunkWithProof(ChunkWithProof),
+}
+
+/// Error returned when constructing an instance of [`ValueOrChunk`].
+#[derive(Debug, Error)]
+pub enum ChunkingError {
+    /// Merkle proof construction error.
+    MerkleConstruction(MerkleConstructionError),
+    /// Serialization error.
+    SerializationError(bytesrepr::Error),
+}
+
+impl Display for ChunkingError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            ChunkingError::MerkleConstruction(error) => f
+                .debug_tuple("MerkleConstructionError")
+                .field(&error)
+                .finish(),
+            ChunkingError::SerializationError(error) => {
+                f.debug_tuple("SerializationError").field(&error).finish()
+            }
+        }
+    }
+}
+
+impl<V> ValueOrChunk<V> {
+    /// Creates an instance of [`ValueOrChunk::Value`] if data size is less than or equal to
+    /// [`ChunkWithProof::CHUNK_SIZE_BYTES`] or a [`ValueOrChunk::ChunkWithProof`] if it is greater.
+    /// In the latter case it will return only the `chunk_index`-th chunk of the value's byte
+    /// representation.
+    pub fn new(data: V, chunk_index: u64) -> Result<Self, ChunkingError>
+    where
+        V: ToBytes,
+    {
+        let bytes = ToBytes::to_bytes(&data).map_err(ChunkingError::SerializationError)?;
+        // NOTE: Cannot accept the chunk size bytes as an argument without changing the
+        // IndexedMerkleProof. The chunk size there is hardcoded and will be used when
+        // determining the chunk.
+        if bytes.len() <= ChunkWithProof::CHUNK_SIZE_BYTES {
```

Comment:
Should we require that `chunk_index` is 0 if that's the case?

Reply:
And fail if it's different? Current approach seems to be to be the most idiot-proof: even if someone passes in an invalid `chunk_index` they will get a correct results. Especially that these chunk sizes are hardcoded so expecting a `chunk_index != 0` where the size of the `bytes` is less than the limit is wrong.

Reply:
> Especially that these chunk sizes are hardcoded so expecting a chunk_index != 0 where the size of the bytes is less than the limit is wrong.
Exactly, which is why returning `Ok` when the chunk index doesn't make sense might hide bugs in other pieces of code. Like if someone writes a loop that goes over multiple indices without checking what the actual count is - they will be under the impression that they got multiple chunks even though they only got one, but repeated multiple times.
I'm tempted to say that this might also potentially be a security issue, but I can't come up with any examples of how it would be, so maybe it isn't. Just feels a bit risky somehow.